PyInstaller/hooks/rthooks/pyi_rth_mplconfig.py

```diff
@@ -10,37 +10,15 @@
 #-----------------------------------------------------------------------------
 
 
-# matplotlib will create $HOME/.matplotlib folder in user's home directory.
-# In this directory there is fontList.cache file which lists paths
-# to matplotlib fonts.
-#
-# When you run your onefile exe for the first time it's extracted to for example
-# "_MEIxxxxx" temp directory and fontList.cache file is created with fonts paths
-# pointing to this directory.
-#
-# Second time you run your exe new directory is created "_MEIyyyyy" but
-# fontList.cache file still points to previous directory which was deleted.
-# And then you will get error like:
-#
-#     RuntimeError: Could not open facefile
-#
-# We need to force matplotlib to recreate config directory every time you run
-# your app.
-
-
-import atexit
 import os
-import shutil
-import tempfile
-
-
-# Put matplot config dir to temp directory.
-configdir = tempfile.mkdtemp()
-os.environ['MPLCONFIGDIR'] = configdir
-
-
-try:
-    # Remove temp directory at application exit and ignore any errors.
-    atexit.register(shutil.rmtree, configdir, ignore_errors=True)
-except OSError:
-    pass
+import sys
+
+path = os.path.join(sys._MEIPASS, 'matplotlib/mpl-cfg')
+if not os.path.exists(path):
+    # In earlier version of matplotlib (3.0.3) the config folder may be empty.
+    # PyInstaller doesn't copy empty folders so in this case
+    # ``matplotlib/mpl-cfg`` will not exist in the PyInstaller build causing
+    # matplotlib to rais a ``FileNotFoundError``.
+    # Ensure the config folder is made (even if it's empty).
+    os.mkdir(path)
+os.environ['MPLCONFIGDIR'] = path
```

Comment:
I think this will cause a regression in a case when a `onedir` build is placed on a read-only filesystem before the first run. And if `fontList.cache` does indeed contain absolute paths referring to `sys._MEIPASS`  as suggested by the original comment, then it will also break if the application is moved after the first run (according to #3959 that might be version-dependent, though).

But overall, does it make sense to collect the config dir ($HOME/.matplotlib) from the build machine in the first place? On a clean build system where matplotlib was never run before, this folder will probably not exist. And when it does exist, `fontList.cache` probably contains the fonts installed on the build system, but not necessary on the target one (the lower part of `fontList.cache` seems to contain system fonts).

I guess we cannot use the default config path because that might introduce conflicts between different built programs with potentially different matplotlib distributions... 

But maybe we could avoid collecting the config dir on build machine, and instead always create and initialize `sys._MEIPASS/matplotlib/mpl-cfg` on the first run (with a fallback to temporary dir if creation fails or if matplotlib version is too old)? This should fix #3959 and prevent regressions in corner cases.

Reply:
Ahh, you'll have to negotiate that with @Legorooj. I think we're getting in matplotlib's way unnecessarly by touching it at all.

Reply:
To be honest we need to discuss this with matplotlib - we're guessing here. Can we merge this for now (and finally put the matplotlib users out of their misery)? Then we can raise this again when we move these hooks into matplotlib's repo.

Reply:
@bwoodsend if you can confirm that the `fontList.cache` issue isn't present in this PR, then I'm happy to merge this. However please be very sure it is/isn't here.

Reply:
Well it's present in onefile mode with matplotlib >= 3.0. But I thought that was the whole point in putting the config in the package - to force it to rebuild the cache which by design is a regression of  #3959?

Reply:
@bwoodsend I'm confused now. Do you mean >= 3.0 or <= 3.0?

Reply:
I mean >= 3.0. Quoting the issue:

> Since matplotlib 3.0 this is no longer the case ([matplotlib/matplotlib#10245](https://github.com/matplotlib/matplotlib/pull/10245)), and the pyinstaller patch causes repeated rebuilds of the font cache, which slows down startup ([matplotlib/matplotlib#13071](https://github.com/matplotlib/matplotlib/issues/13071)).

We're not creating a tempdir using `tempdir` but as an unpacked onefile is a temporary dir anyway it has the same effect.

Reply:
@Legorooj prod prod...

Reply:
Alright I'm still confused. 

> Well it's present in onefile mode with matplotlib >= 3.0.

You appear to be saying that the issue with `fontList.cache` having absolute paths still exists when using the updated hooks in this PR. Am I interpreting this correctly, or is the issue fixed in this PR and I'm just confused?

Reply:
#3959 isn't about having absolute paths - it's about the performance cost of having to rebuild the cache every run (possibly the source of your confusion?).

By copying the config dir (including `fontList.cache`) into the build this PR introduces absolute paths which shouldn't be there. And it still doesn't fix #3959 because, by putting cached contents into our temporary onefile build, any caching is lost. 

Again, I really don't think we should be doing this... I'd much rather revert the 2nd commit and merge the 1st. And possibly ditch our runtime hook which would fix #3559.

Reply:
Alright, drop the second commit, and I'll merge this.